### PR TITLE
METAL: memory efficient, crash free and kernel prefix fix in include …

### PR DESCRIPTION
**Pull request by Maroc-OS**
... include metal library

using newBufferWithBytesNoCopy_length_options_deallocator_ for the new buffer, fix the Metal crash's with proper metal purgeable method, and delete a small delimiter ';' from kernel prefix in MetalCodegen class this was a typo somehow

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/runtime/ops_metal.py             19              13
------------------------------------------------------------
lines added in the tinygrad folder: 6
```

**Comment by geohot** [comment]
Will merge after there's Metal CI tests!

File: tinygrad/runtime/ops_metal.py [hunk 1]
```diff
@@ -10,32 +9,37 @@
...
-  def __init__(self, size:int, dtype:DType): super().__init__(size, dtype, METAL.device.newBufferWithLength_options_(size*dtype.itemsize, Metal.MTLResourceStorageModeShared))
-  def __del__(self):
-    self._buf.release()
+  def __init__(self, size:int, dtype:DType) -> None:
+    super().__init__(size, dtype, METAL.device.newBufferWithBytesNoCopy_length_options_deallocator_(None, size * dtype.itemsize, Metal.MTLResourceStorageModeShared, None))
+  def __del__(self) -> None:
+    self.release()
     super().__del__()
+  def release(self) -> None:
+    purgeable_state = 2  # 2 represents the MTLPurgeableState.Empty state
```
Comment (kposborne2) [inline tinygrad/runtime/ops_metal.py:29]: On this webapge, it says that "2" is "nonvolatile", is it different through the py-objc framework? https://developer.apple.com/documentation/metal/mtlpurgeablestate/mtlpurgeablestatenonvolatile

File: tinygrad/runtime/ops_metal.py [hunk 2]
```diff
@@ -10,32 +9,37 @@
...
-  def __del__(self):
-    self._buf.release()
+  def __init__(self, size:int, dtype:DType) -> None:
+    super().__init__(size, dtype, METAL.device.newBufferWithBytesNoCopy_length_options_deallocator_(None, size * dtype.itemsize, Metal.MTLResourceStorageModeShared, None))
+  def __del__(self) -> None:
+    self.release()
     super().__del__()
+  def release(self) -> None:
+    if self._buf:
+      self._buf.setPurgeableState_(Metal.MTLPurgeableStateEmpty)
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:30]: What does this do?

Comment (geohot) [reply tinygrad/runtime/ops_metal.py:30]: release alone isn't enough?

File: tinygrad/runtime/ops_metal.py [hunk 3]
```diff
@@ -10,32 +9,37 @@
...
     for cbuf in self.mtl_buffers_in_flight: cbuf.waitUntilCompleted()
     self.mtl_buffers_in_flight.clear()
 METAL = _METAL()
 
 class RawMetalBuffer(RawBufferMapped):
-  def __init__(self, size:int, dtype:DType): super().__init__(size, dtype, METAL.device.newBufferWithLength_options_(size*dtype.itemsize, Metal.MTLResourceStorageModeShared))
-  def __del__(self):
-    self._buf.release()
+  def __init__(self, size:int, dtype:DType) -> None:
+    super().__init__(size, dtype, METAL.device.newBufferWithBytesNoCopy_length_options_deallocator_(None, size * dtype.itemsize, Metal.MTLResourceStorageModeShared, None))
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:24]: Why is this constructor better?

File: tinygrad/runtime/ops_metal.py [hunk 4]
```diff
@@ -45,42 +49,44 @@ def __init__(self, name:str, prg:str):
     else:
       options = Metal.MTLCompileOptions.alloc().init()
       self.library = unwrap(METAL.device.newLibraryWithSource_options_error_(prg, options, None))
+    if self.library is None:
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:52]: How will this ever happen?

**Comment by geohot** [comment]
There's functional changes in here mixed with others, which makes it hard to read and review.

* add types (fine)
* rename variables (why?)
* small functionality changes (separate these)

**Comment by geohot** [comment]
Yea needs to be separate PRs. I can't read this diff.

**Pull request closed without merging**